File: src-tauri/src/commands/connection.rs
use crate::sidecar::SidecarManager;
use serde_json::Value;

#[tauri::command]
pub async fn connection_list(sidecar: tauri::State<'_, SidecarManager>) -> Result<String, String> {
    let result = sidecar.send_request("connection.list", None).await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

#[tauri::command]
pub async fn connection_get(
    sidecar: tauri::State<'_, SidecarManager>,
    id: String,
) -> Result<String, String> {
    let params = serde_json::json!({ "id": id });
    let result = sidecar.send_request("connection.get", Some(params)).await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

#[tauri::command]
pub async fn connection_save(
    sidecar: tauri::State<'_, SidecarManager>,
    connection: Value,
    password: Option<String>,
) -> Result<String, String> {
    let params = serde_json::json!({
        "connection": connection,
        "password": password
    });
    let result = sidecar.send_request("connection.save", Some(params)).await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

#[tauri::command]
pub async fn connection_delete(
    sidecar: tauri::State<'_, SidecarManager>,
    id: String,
) -> Result<String, String> {
    let params = serde_json::json!({ "id": id });
    let result = sidecar
        .send_request("connection.delete", Some(params))
        .await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

#[tauri::command]
pub async fn connection_test(
    sidecar: tauri::State<'_, SidecarManager>,
    connection: Value,
    password: Option<String>,
) -> Result<String, String> {
    let params = serde_json::json!({
        "connection": connection,
        "password": password
    });
    let result = sidecar.send_request("connection.test", Some(params)).await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

#[tauri::command]
pub async fn connection_connect(
    sidecar: tauri::State<'_, SidecarManager>,
    id: String,
) -> Result<String, String> {
    let params = serde_json::json!({ "id": id });
    let result = sidecar
        .send_request("connection.connect", Some(params))
        .await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

#[tauri::command]
pub async fn connection_disconnect(
    sidecar: tauri::State<'_, SidecarManager>,
    id: String,
) -> Result<String, String> {
    let params = serde_json::json!({ "id": id });
    let result = sidecar
        .send_request("connection.disconnect", Some(params))
        .await?;
    Ok(serde_json::to_string(&result).map_err(|e| e.to_string())?)
}

File: src/commands/connection.ts
import { invoke } from "@tauri-apps/api/core";

export type AuthType = "SqlAuth" | "ConnectionString" | "EntraMfa";
export type EncryptMode = "Mandatory" | "Optional" | "Strict";

export interface ConnectionInfo {
  id: string;
  name: string;
  serverName: string;
  authType: AuthType;
  username?: string;
  credentialRef?: string;
  database?: string;
  encrypt: EncryptMode;
  trustServerCertificate: boolean;
  connectionString?: string;
  color?: string;
  lastUsed?: string;
  createdAt: string;
}

export interface ConnectionTestResult {
  success: boolean;
  error?: string;
}

export interface ConnectionConnectResult {
  connectionId: string;
}

export interface ConnectionDeleteResult {
  deleted: boolean;
}

export async function connectionList(): Promise<ConnectionInfo[]> {
  const result = await invoke<string>("connection_list");
  return JSON.parse(result);
}

export async function connectionGet(id: string): Promise<ConnectionInfo | null> {
  const result = await invoke<string>("connection_get", { id });
  return JSON.parse(result);
}

export async function connectionSave(
  connection: ConnectionInfo,
  password?: string,
  clearCredential?: boolean
): Promise<ConnectionInfo> {
  const result = await invoke<string>("connection_save", { connection, password, clearCredential: clearCredential ?? false });
  return JSON.parse(result);
}

export async function connectionDelete(id: string): Promise<ConnectionDeleteResult> {
  const result = await invoke<string>("connection_delete", { id });
  return JSON.parse(result);
}

export async function connectionTest(
  connection: ConnectionInfo,
  password?: string
): Promise<ConnectionTestResult> {
  const result = await invoke<string>("connection_test", { connection, password });
  return JSON.parse(result);
}

export async function connectionConnect(id: string): Promise<ConnectionConnectResult> {
  const result = await invoke<string>("connection_connect", { id });
  return JSON.parse(result);
}

export async function connectionDisconnect(id: string): Promise<void> {
  await invoke<string>("connection_disconnect", { id });
}
